### PR TITLE
GH-10787 UserIdService to determine correct cart user

### LINKED ISSUE #10787: [Major] Cart integration does not work in ASM scenario
The product configurator scenarios don't work in the ASM scenario for 2 reasons:

- SPA: In case user is logged in, the actual cart user is not properly sent to the backend (for that we need to make use of  userIdService.getUserId()
- Commerce backend: Some security related access checks fail, because they don't allow the agent user to retrieve a configuration. That's valid for the Variant Configurator only.

How to reproduce: Login as asagent, start a session for and end-user. The addToCart process fails with a 400'er exception from the backend.

If this is fixed, the ASM scenario works fine for the textfield configurator

### PR DESCRIPTION
**Pull request by ChristophHi**

### CODE REVIEW

File: feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts [hunk 1]
```diff
@@ -65,18 +68,23 @@ export class CommonConfiguratorUtilsService {
...
+  getCartId(cart: Cart): string {
     return cart.user.uid === OCC_USER_ID_ANONYMOUS ? cart.guid : cart.code;
   }
+
   /**
-   * Gets cart user
+   * Gets user assigned to a cart which we can use for OCC requests
    * @param cart Cart
-   * @returns User identifier
+   * @returns Observable of user identifiers
```
Comment (schuppell) [inline feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts]: Is it correct that the observable contains user identifiers not just one identifier?

File: feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts [hunk 2]
```diff
@@ -56,21 +56,24 @@ export class ConfiguratorCartService {
...
-              };
-              this.store.dispatch(
-                new ConfiguratorActions.ReadCartEntryConfiguration(
-                  readFromCartEntryParameters
-                )
-              );
+              this.commonConfigUtilsService
+                .getUserId(cartState.value)
+                .pipe(take(1))
+                .subscribe((id) => {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts]: I would suggest to call it userId to make it more readable

File: feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts [hunk 3]
```diff
@@ -136,17 +138,22 @@ export class ConfiguratorCartService {
...
-          configId: configId,
-          owner: owner,
-        };
-        this.store.dispatch(
-          new ConfiguratorActions.AddToCart(addToCartParameters)
-        );
+        this.commonConfigUtilsService
+          .getUserId(cartState.value)
+          .pipe(take(1))
+          .subscribe((id) => {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts]: I would suggest to call it userId to make it more readable

File: feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts [hunk 4]
```diff
@@ -162,17 +169,24 @@ export class ConfiguratorCartService {
...
-        const parameters: Configurator.UpdateConfigurationForCartEntryParameters = {
-          userId: this.commonConfigUtilsService.getUserId(cartState.value),
-          cartId: cartId,
-          cartEntryNumber: configuration.owner.id,
-          configuration: configuration,
-        };
+        this.commonConfigUtilsService
+          .getUserId(cartState.value)
+          .pipe(take(1))
+          .subscribe((id) => {
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts]: I would suggest to call it userId to make it more readable

File: feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts [hunk 5]
```diff
@@ -162,17 +169,24 @@ export class ConfiguratorCartService {
...
+            const cartId = this.commonConfigUtilsService.getCartId(
+              cartState.value
+            );
+            const parameters: Configurator.UpdateConfigurationForCartEntryParameters = {
+              userId: id,
+              cartId: cartId,
```
Comment (schuppell) [inline feature-libs/product-configurator/rulebased/core/facade/configurator-cart.service.ts]: In all other methods cartId is not a variable an will be filled directly from the method return value. Is there a reason to first create a const cartId?

File: feature-libs/product-configurator/textfield/core/facade/configurator-textfield.service.ts [hunk 6]
```diff
@@ -92,16 +92,21 @@ export class ConfiguratorTextfieldService {
...
-          configuration: configuration,
-          quantity: 1,
-        };
-        this.store.dispatch(
-          new ConfiguratorTextfieldActions.AddToCart(addToCartParameters)
-        );
+        this.configuratorUtils
+          .getUserId(cartState.value)
+          .pipe(take(1))
+          .subscribe((id) => {
```
Comment (schuppell) [inline feature-libs/product-configurator/textfield/core/facade/configurator-textfield.service.ts]: I would suggest to call it userId to make it more readable

File: feature-libs/product-configurator/textfield/core/facade/configurator-textfield.service.ts [hunk 7]
```diff
@@ -119,17 +124,22 @@ export class ConfiguratorTextfieldService {
...
-        };
-        this.store.dispatch(
-          new ConfiguratorTextfieldActions.UpdateCartEntryConfiguration(
-            updateCartParameters
-          )
-        );
+        this.configuratorUtils
+          .getUserId(cartState.value)
+          .pipe(take(1))
+          .subscribe((id) => {
```
Comment (schuppell) [inline feature-libs/product-configurator/textfield/core/facade/configurator-textfield.service.ts]: I would suggest to call it userId to make it more readable

File: feature-libs/product-configurator/textfield/core/facade/configurator-textfield.service.ts [hunk 8]
```diff
@@ -144,25 +154,37 @@ export class ConfiguratorTextfieldService {
...
-      tap((readFromCartEntryParameters) =>
-        this.store.dispatch(
-          new ConfiguratorTextfieldActions.ReadCartEntryConfiguration(
-            readFromCartEntryParameters
+      switchMap((cartState) =>
+        this.configuratorUtils
+          .getUserId(cartState.value)
+          .pipe(
+            take(1),
+            map((id) => ({ cartState, id }))
```
Comment (schuppell) [inline feature-libs/product-configurator/textfield/core/facade/configurator-textfield.service.ts]: I would suggest to call it userId to make it more readable

**Review by schuppell — CHANGES_REQUESTED** [review]
Overall looks good, I would suggest to rename a few variables to make them more clear.

**Review by schuppell — DISMISSED** [review]

File: feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts [hunk 9]
```diff
@@ -65,18 +68,23 @@ export class CommonConfiguratorUtilsService {
...
-  public getUserId(cart: Cart): string {
-    return cart.user.uid === OCC_USER_ID_ANONYMOUS
-      ? cart.user.uid
-      : OCC_USER_ID_CURRENT;
+  getUserId(cart: Cart): Observable<string> {
+    return this.userIdService
+      .getUserId()
+      .pipe(
+        map((id) =>
+          cart.user.uid === OCC_USER_ID_ANONYMOUS ? cart.user.uid : id
```
Comment (Platonn) [inline feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts]: So far, we used the `user` embedded in the `cart` only for getting `email` in the _guest checkout_ functionality (when there was no real account).

Why we need to use `user` embedded in the `cart` object in _product configurator_ feature? Can't we use simply rely on the value from `UserIdService`?

Comment (ChristophHi) [reply feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts]: Agree, not needed. We can use userIdService directly and can get rid of this method

**Review by schuppell — APPROVED** [review]

**Pull request merged**